### PR TITLE
WIP: Begin Introduction of new 'sunset' API-Resource annotation in order t…

### PR DESCRIPTION
**Pull request by blanky0230**
…o support: sunset header
I'm not quite sure this works as excpected :) May someone please take a look?

I'm also rahter new to behat. I suppose I'd need to create some Behat Tests for this as well.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Related to | #2245 
| License       | MIT
| Doc PR        | //TODO!

*Please update this template with something that matches your PR*

### CODE REVIEW

File: src/Annotation/ApiResource.php [hunk 1]
```diff
@@ -52,7 +52,8 @@
  *     @Attribute("shortName", type="string"),
  *     @Attribute("subresourceOperations", type="array"),
  *     @Attribute("swaggerContext", type="array"),
- *     @Attribute("validationGroups", type="mixed")
+ *     @Attribute("validationGroups", type="mixed"),
+ *     @Attribute("sunset", type="string")
```
Comment (dunglas) [inline src/Annotation/ApiResource.php]: Can you move it after `subresourceOperations`? (alpha order)

File: src/Metadata/Resource/ResourceMetadata.php [hunk 2]
```diff
@@ -41,6 +42,7 @@ public function __construct(string $shortName = null, string $description = null
         $this->subresourceOperations = $subresourceOperations;
         $this->graphql = $graphql;
         $this->attributes = $attributes;
+        $this->sunset = $sunset;
```
Comment (dunglas) [inline src/Metadata/Resource/ResourceMetadata.php]: This is not necessary IMO, an attribute would be better suited.

File: src/Sunset/EventListener/AddSunsetHeaderListener.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+/**
```
Comment (dunglas) [inline src/Sunset/EventListener/AddSunsetHeaderListener.php]: To be removed

File: src/Sunset/EventListener/AddSunsetHeaderListener.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace ApiPlatform\Core\Sunset\EventListener;
+
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+
+class AddSunsetHeaderListener
+{
+    private $sunset;
+
+    public function __construct(string $sunset = null)
```
Comment (dunglas) [inline src/Sunset/EventListener/AddSunsetHeaderListener.php]: The listener should be stateless. You should detect the current resource using `RequestAttributesExtractor` and call `ResourceMetadataFactory` instead.

File: tests/Sunset/AddSunsetHeaderListenerTest.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+        $response = new Response();
+
+        $event = $this->prophesize(FilterResponseEvent::class);
+        $event->getResponse()->willReturn($response)->shouldBeCalled();
+
+        $listener = new AddSunsetHeaderListener('2018-10-10');
+        $listener->onKernelResponse($event->reveal());
+
+        $this->assertEquals('2018-10-10', $response->headers->get('Sunset'));
```
Comment (dunglas) [inline tests/Sunset/AddSunsetHeaderListenerTest.php]: The data must follow the HTTP date format: https://tools.ietf.org/id/draft-wilde-sunset-header-03.html#sunset-header. You should call `\DateTimeImmutable($sunsetAttributeValue)` then call `format` with the appropriate value (see https://stackoverflow.com/questions/21120882/the-date-time-format-used-in-http-headers).

Comment (blanky0230) [reply tests/Sunset/AddSunsetHeaderListenerTest.php]: Thanks! I had some trouble figuring out what format to use and how to enforce it!

File: src/Annotation/ApiResource.php (hunk 1 above)
Comment (blanky0230) [reply src/Annotation/ApiResource.php]: should now be the case

File: src/Metadata/Resource/ResourceMetadata.php (hunk 2 above)
Comment (blanky0230) [reply src/Metadata/Resource/ResourceMetadata.php]: removed

File: src/Sunset/EventListener/AddSunsetHeaderListener.php (hunk 3 above)
Comment (blanky0230) [reply src/Sunset/EventListener/AddSunsetHeaderListener.php]: Gotta set up my IDE :/ Sorry.

File: src/Sunset/EventListener/AddSunsetHeaderListener.php (hunk 4 above)
Comment (blanky0230) [reply src/Sunset/EventListener/AddSunsetHeaderListener.php]: Made some changes. I am still not 100% confident this is how it should be done though. Do you think there should also be some sort of enforcing the Date format?

**Pull request closed without merging**